File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/calculo/juros/ApuracaoDeJurosOptimizerListSearch.java
/*
 * Decompiled with CFR 0.152.
 */
package br.jus.trt8.pjecalc.negocio.dominio.calculo.juros;

import br.jus.trt8.pjecalc.base.comum.Competencia;
import br.jus.trt8.pjecalc.negocio.comum.OcorrenciaIterator;
import br.jus.trt8.pjecalc.negocio.comum.OptimizerListSearch;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.juros.ApuracaoDeJuros;
import java.util.Collection;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

public class ApuracaoDeJurosOptimizerListSearch
extends OptimizerListSearch<Competencia, ApuracaoDeJuros> {
    private Map<Competencia, OcorrenciaIterator<ApuracaoDeJuros>> map;

    @Override
    public OptimizerListSearch<Competencia, ApuracaoDeJuros> init(Collection<ApuracaoDeJuros> collection) {
        this.map = new HashMap<Competencia, OcorrenciaIterator<ApuracaoDeJuros>>();
        for (ApuracaoDeJuros ocorrencia : collection) {
            Competencia competencia = Competencia.getInstance(ocorrencia.getCompetencia());
            OcorrenciaIterator<ApuracaoDeJuros> iterator = this.map.get(competencia);
            if (iterator == null) {
                iterator = new OcorrenciaIterator<ApuracaoDeJuros>(ocorrencia);
                this.map.put(competencia, iterator);
                continue;
            }
            iterator.add(ocorrencia);
        }
        return this;
    }

    @Override
    public Iterator<ApuracaoDeJuros> search(Competencia key) {
        OcorrenciaIterator<ApuracaoDeJuros> iterator = this.map.get(key);
        if (iterator != null) {
            iterator.gotoFirst();
        }
        return iterator;
    }
}

File: src/lib/pjecalc/core/dominio/calculo/juros/apuracao-de-juros-optimizer-list-search.ts
/**
 * PJe-Calc v2.15.1 — ApuracaoDeJurosOptimizerListSearch
 * Porte 1:1 de: br.jus.trt8.pjecalc.negocio.dominio.calculo.juros.ApuracaoDeJurosOptimizerListSearch
 *
 * Ref Java: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/calculo/juros/ApuracaoDeJurosOptimizerListSearch.java
 *
 * Indexa ApuracaoDeJuros por Competencia (mes/ano) para acesso O(1) em loops
 * de cálculo de juros.
 */
import { Competencia } from '../../../base/comum/competencia';
import { OcorrenciaIterator, OptimizerListSearch } from '../../../comum/optimizer-list-search';
import type { ApuracaoDeJuros } from './apuracao-de-juros';

export class ApuracaoDeJurosOptimizerListSearch extends OptimizerListSearch<Competencia, ApuracaoDeJuros> {
  private map: Map<string, OcorrenciaIterator<ApuracaoDeJuros>> = new Map();

  init(collection: Iterable<ApuracaoDeJuros>): this {
    this.map = new Map();
    for (const ocorrencia of collection) {
      const comp = ocorrencia.getCompetencia();
      if (!comp) continue;
      const key = Competencia.getInstance(comp).toKey();
      const iterator = this.map.get(key);
      if (!iterator) {
        this.map.set(key, new OcorrenciaIterator<ApuracaoDeJuros>(ocorrencia));
      } else {
        iterator.add(ocorrencia);
      }
    }
    return this;
  }

  search(key: Competencia): OcorrenciaIterator<ApuracaoDeJuros> | null {
    const k = key.toKey();
    const iterator = this.map.get(k) ?? null;
    if (iterator) iterator.gotoFirst();
    return iterator;
  }
}
